apiserver/common/networkingcommon/networkconfigapi.go

```diff
@@ -91,33 +91,40 @@ func (api *NetworkConfigAPI) SetObservedNetworkConfig(args params.SetMachineNetw
 	return errors.Trace(api.st.ApplyOperation(api.getModelOp(m, devs)))
 }
 
-// fixUpFanSubnets takes network config and updates
-// Fan subnets with proper CIDR.
-// See network/fan.go for more detail on how Fan overlays
-// are divided into segments.
+// fixUpFanSubnets takes network config and updates any addresses in Fan
+// networks with the CIDR of the zone-specific segments. We need to do this
+// because they are detected on-machine as being in the /8 overlay subnet,
+// which is a superset of the zone segments.
+// See core/network/fan.go for more detail on how Fan overlays are divided
+// into segments.
 func (api *NetworkConfigAPI) fixUpFanSubnets(networkConfig []params.NetworkConfig) ([]params.NetworkConfig, error) {
 	subnets, err := api.st.AllSubnetInfos()
 	if err != nil {
 		return nil, errors.Trace(err)
 	}
 
-	var fanSubnets network.SubnetInfos
+	fanSubnets := make(map[string]*net.IPNet)
 	for _, subnet := range subnets {
 		if subnet.FanOverlay() != "" {
-			fanSubnets = append(fanSubnets, subnet)
-		}
-	}
-	for i := range networkConfig {
-		localIP := net.ParseIP(networkConfig[i].Address)
-
-		for _, sub := range fanSubnets {
-			fanNet, err := sub.ParsedCIDRNetwork()
+			fanSub, err := subnet.ParsedCIDRNetwork()
 			if err != nil {
 				return nil, errors.Trace(err)
 			}
-			if fanNet != nil && fanNet.Contains(localIP) {
-				networkConfig[i].CIDR = sub.CIDR
-				break
+			fanSubnets[subnet.CIDR] = fanSub
+		}
+	}
+
+	if len(fanSubnets) == 0 {
+		return networkConfig, nil
+	}
+
+	for i := range networkConfig {
+		for j := range networkConfig[i].Addresses {
+			for cidr, fanNet := range fanSubnets {
+				if fanNet != nil && fanNet.Contains(net.ParseIP(networkConfig[i].Addresses[j].Value)) {
```

Comment:
Is it actually possible for fanNet to be nil? I don't mind the caution, but given you just populate the map a few lines above, I wanted to make sure I understood the algorithm correctly. Is it just that ParsedCIDRNetwork might return nil for the subnet but not give an error?
Or is this just a belt-and-suspenders check to make sure we don't have a panic where it would be stupid to not just skip the item.

Given the nesting, this might be clearer as:
```
for i, cfg := range networkConfig {
	for j, addr := range cfg.Addresses {
          ip := net.ParseIP(addr.Value)
		for cidr, fanNet := range fanSubnets {
    if fanNet == nil {
      continue
    }
    if fanNet.Contains(ip) {
      // update the original maps rather than the local variables
      networkConfig[i].Addresses[j].CIDR = cidr
      break
    }
...
```

Reply:
(it also led me to notice that you're parsing the Address for every fanSubnet, which can be pulled out of the inner loop)

Reply:
I pushed a small tweak to pull out the address parsing to the top of the loop